### PR TITLE
Added GoogleNet (Inception V1) Model

### PR DESCRIPTION
**Pull request by Sarvesh-Kesharwani**
Corresponding to: #38

### CODE REVIEW

**Comment by Sarvesh-Kesharwani** [comment]
chrome_tocLYwuZ9V

**Comment by Sarvesh-Kesharwani** [comment]
chrome_CFz5HWx7ty

File: ivy_models/googlenet/googlenet.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# global
+import ivy
+import ivy_models
+from .layers import ConvBlock, Inception, Auxiliary
+
+
+class GoogLeNet(ivy.Module):
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:7]: please use the ivy `BaseModel` class here

File: ivy_models/googlenet/googlenet.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    Args::
+        num_classes (int): Number of output classes. Defaults to 1000.
+        base_width (int): The base width of the ResNet. Defaults to 64.
+        v (ivy.Container): Unused parameter. Can be ignored.
+
+    """
+
+    def __init__(
+        self,
+        num_classes=1000,
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:20]: we aim to support dual format and spec or keyword arg initialisation, please also add them above, thanks!

File: ivy_models/googlenet/googlenet.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        v: ivy.Container = None,
+    ):
+        if v is not None:
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:23]: this shouldn't be needed as we already have `v=v` in `super().__init__()` below

File: ivy_models/googlenet/googlenet.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        self.pool5 = ivy.MaxPool2D([3, 3], 2, "SAME")
+
+        self.inception5A = Inception(832, 256, 160, 320, 32, 128, 128)
+        self.inception5B = Inception(832, 384, 192, 384, 48, 128, 128)
+        self.pool6 = ivy.AdaptiveAvgPool2d([1, 1])
+
+        self.dropout = ivy.Dropout(0.4)
+        self.fc = ivy.Linear(1024, self.num_classes)
+
+    def _forward(self, x):
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:60]: once we have the dual format support, the forward should have an extra condition block, please ref alexnet model

File: ivy_models/googlenet/googlenet.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    W_KEY = ["conv/weight"]
+    new_mapping = new_key
+    if any([kc in old_key for kc in W_KEY]):
+        new_mapping = {"key_chain": new_key, "pattern": "b c h w -> h w c b"}
+    return new_mapping
+
+
+def inceptionNet_v1(pretrained=True):
+    """InceptionNet-V1 model"""
+    if not pretrained:
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:104]: just a small refactor here, as we would want to avoid double model initialisation for better mem usage and efficiency, please ref resnet model for the weights loading code

File: ivy_models_tests/googlenet/test_googlenet.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import os
+import ivy
+import pytest
+import numpy as np
+
+from ivy_models.googlenet import inceptionNet_v1
+from ivy_models_tests import helpers
+
+
+def np_softmax(inputs):
```
Comment (juliagsy) [inline ivy_models_tests/googlenet/test_googlenet.py:10]: there is an equivalent helper in the `ivy_models_tests/helpers` folder

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey Sarvesh! Thanks for the great work! I've left some comments below mainly related to support general additional support and refactor. The updated guidelines can also be found here: https://docs.google.com/document/d/1NfVngAQlQQlxLGEg47O2lnfn60M5xZekZFei9k5RJd8/edit?usp=sharing
Please feel free to lmk if you have any questions! Thanks!

File: ivy_models/googlenet/layers.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
+import ivy
+
+
+class ConvBlock(ivy.Module):
```
Comment (juliagsy) [inline ivy_models/googlenet/layers.py:4]: a small request, would you mind renaming the layers here to `GoogleNet<Layer>` for convention purposes? thanks!

File: ivy_models_tests/googlenet/test_googlenet.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+jax.config.update("jax_enable_x64", False)
+
+load_weights = random.choice([False, True])
+model = inceptionNet_v1(pretrained=load_weights)
+v = ivy.to_numpy(model.v)
+
+
+@pytest.mark.parametrize("data_format", ["NHWC", "NCHW"])
+def test_alexnet_tiny_img_classification(device, f, fw, data_format):
```
Comment (juliagsy) [inline ivy_models_tests/googlenet/test_googlenet.py:20]: testname typo

File: ivy_models/googlenet/googlenet.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        self.pool1 = ivy.MaxPool2D([3, 3], 2, "SAME")
+
+        self.conv2 = ConvBlock(64, 64, [1, 1], 1, padding="SAME")
+        self.conv3 = ConvBlock(64, 192, [3, 3], 1, padding="SAME")
+        # self.pool3 = ivy.MaxPool2D([3, 3], 2, "SAME", data_format="NCHW")
+        self.pool3 = ivy.MaxPool2D([3, 3], 2, "SAME")
+
+        self.inception3A = Inception(192, 64, 96, 128, 16, 32, 32)
+        self.inception3B = Inception(256, 128, 128, 192, 32, 96, 64)
+        # self.pool4 = ivy.MaxPool2D([3, 3], 2, "SAME", data_format="NCHW")
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:63]: please also clean up code

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey Sarvesh! Thanks for the great work! The code looks good to me overall and we're good to merge once the final comments are addressed and that the tests are passing! Thanks!

File: ivy_models_tests/googlenet/test_googlenet.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    """Test AlexNet image classification."""
```
Comment (a0m0rajab) [inline ivy_models_tests/googlenet/test_googlenet.py:21]:
```suggestion
    """Test GoogleNet image classification."""
```

Just adding my two cents :)

File: ivy_models/googlenet/googlenet.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    W_KEY = ["conv/weight"]
+    new_mapping = new_key
+    if any([kc in old_key for kc in W_KEY]):
+        new_mapping = {"key_chain": new_key, "pattern": "b c h w -> h w c b"}
+    return new_mapping
+
+
+def inceptionNet_v1(pretrained=True, num_classes=1000, dropout=0, data_format="NHWC"):
+    """InceptionNet-V1 model"""
+    model = GoogLeNet(num_classes=num_classes, dropout=dropout, data_format=data_format)
```
Comment (juliagsy) [inline ivy_models/googlenet/googlenet.py:138]: aux_dropout

File: ivy_models_tests/googlenet/test_googlenet.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    model.v = ivy.asarray(v)
+    logits, _, _ = model(img, data_format=data_format)
+
+    # Cardinality test
+    assert logits.shape == tuple([ivy.to_scalar(batch_shape), num_classes])
+
+    # Value test
+    if load_weights:
+        np_out = ivy.to_numpy(logits[0])
+        true_indices = np.array([282, 281, 285])
```
Comment (juliagsy) [inline ivy_models_tests/googlenet/test_googlenet.py:44]: this would then be `true_indices = np.array([281, 285, 282])`

File: ivy_models_tests/googlenet/test_googlenet.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        calc_indices = np.argsort(np_out)[-5:][::-1]
```
Comment (juliagsy) [inline ivy_models_tests/googlenet/test_googlenet.py:45]: -3 instead of -5 here

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey Sarvesh! All dropout values should be 0 including aux_dropout. I've left some comments, once they are all in place, the test should pass for both load weights = True and False (I've tried it out :D) Thanks!
PS: Please also clean up the code after that and we'll then be able to look into merging, thanks!

**Review by juliagsy — COMMENTED** [review]
Hey Sarvesh! Thanks for the changes! The PR may not be ready yet, but just thought I'd leave a comment before I forget, please remove any changes unrelated to models and keep googlenet related updates only, thanks!

**Comment by Sarvesh-Kesharwani** [comment]
> Hey Sarvesh! Thanks for the great work! The code looks good to me overall and we're good to merge once the final comments are addressed and that the tests are passing! Thanks! PS: please use "squash and merge" when merging :D

Sure will take care from next time.

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey Sarvesh! I was about to merge it and doing a final check, but the pretrained weights loading result is not correct anymore. I'm not sure what changed (most likely the architecture and its config) but could you please double check (one thing is that I had to make `fc` with_bias=True for the weights mapping to map correctly but I didn't have to change anything previously to make the tests run)
I'm also pretty sure that during the last review, tests were passing with minimal change (the review comments), let me know once it's done, thanks!

**Comment by juliagsy** [comment]
Hey! A quick update, seems like inception3 had a similar issue 🤔
some resnet tests are not passing anymore now which should be, could you try looking into this direction? maybe try to see if any module changed within ivy or if they have shared layers which may have some implementation change? thanks!

**Comment by Sarvesh-Kesharwani** [comment]
> Hey! A quick update, seems like inception3 had a similar issue 🤔 some resnet tests are not passing anymore now which should be, could you try looking into this direction? maybe try to see if any module changed within ivy or if they have shared layers which may have some implementation change? thanks!

image

**Comment by Sarvesh-Kesharwani** [comment]
> Hey! A quick update, seems like inception3 had a similar issue 🤔 some resnet tests are not passing anymore now which should be, could you try looking into this direction? maybe try to see if any module changed within ivy or if they have shared layers which may have some implementation change? thanks!

Working on the resnet failure issue.

**Review by juliagsy — APPROVED** [review]
I'm temp commenting out logins value tests (the float dtype ones)
otherwise lgtm, thanks!

**Review by juliagsy — APPROVED** [review]
lgtm, thanks!
